prjxray/util.py

```diff
@@ -172,43 +172,34 @@ def addr2btype(base_addr):
     return block_type_i2s[block_type_i]
 
 
-def gen_tile_bits(db_root, tilej, strict=False, verbose=False):
+def gen_tile_bits(tile_segbits, tile, strict=False, verbose=False):
     '''
-    For given tile yield
+    For given tile and corresponding db_file structure yield
     (absolute address, absolute FDRI bit offset, tag)
 
-    For each address space
-    Find applicable files
-    For each tag bit in those files, calculate absolute address and bit offsets
+    db_file:
+    {'<block_type_1>': [<bit lines>], '<block_type_2>': [<bit lines>], ...}
 
-    Sample file names:
-    segbits_clbll_l.db
-    segbits_int_l.db
-    segbits_bram_l.block_ram.db
+    For each tag bit in the corresponding block_type entry, calculate absolute address and bit offsets
     '''
-    for block_type, blockj in tilej["bits"].items():
-        baseaddr = int(blockj["baseaddr"], 0)
-        bitbase = 32 * blockj["offset"]
-
-        if block_type == "CLB_IO_CLK":
-            fn = "%s/segbits_%s.db" % (db_root, tilej["type"].lower())
-        else:
-            fn = "%s/segbits_%s.db.%s" % (
-                db_root, tilej["type"].lower(), block_type.lower())
-        # tilegrid runs a lot earlier than fuzzers
-        # may not have been created yet
-        verbose and print("Check %s: %s" % (fn, os.path.exists(fn)))
-        if strict:
-            assert os.path.exists(fn)
-        elif not os.path.exists(fn):
-            continue
-
-        for line, (tag, bits, mode) in parse_db_lines(fn):
-            assert mode is None
-            for bitstr in bits:
+
+    for block_type in tile_segbits:
+        assert block_type.value in tile[
+            "bits"], "block type %s is not present in current tile" % block_type.value
+
+        block = tile["bits"][block_type.value]
+
+        baseaddr = int(block["baseaddr"], 0)
+        bitbase = 32 * block["offset"]
+        frames = block["frames"]
+
+        for tag in tile_segbits[block_type]:
+            for bit in tile_segbits[block_type][tag]:
                 # 31_06
-                _bit_inv, (bit_addroff, bit_bitoff) = parse_tagbit(bitstr)
-                yield (baseaddr + bit_addroff, bitbase + bit_bitoff, tag)
+                word_column, word_bit, isset = bit
```

Comment:
Please used the named fields rather than expanding the tuple

Reply:
Done